modules/460-log-shipper/hooks/internal/handler/generator.go

```diff
@@ -0,0 +1,183 @@
+/*
+Copyright 2021 Flant JSC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package handler
+
+import (
+	"github.com/flant/addon-operator/pkg/module_manager/go_hook"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/deckhouse/deckhouse/modules/460-log-shipper/hooks/internal/v1alpha1"
+	"github.com/deckhouse/deckhouse/modules/460-log-shipper/hooks/internal/vector/config"
+	"github.com/deckhouse/deckhouse/modules/460-log-shipper/hooks/internal/vector/model"
+)
+
+type Generator struct {
+	logDestinations map[string]v1alpha1.ClusterLogDestination
+}
+
+func FromInput(input *go_hook.HookInput) *Generator {
+	destNameToObject := make(map[string]v1alpha1.ClusterLogDestination)
+	snap := input.Snapshots["cluster_log_destination"]
+
+	// load all destinations into the destination map
+	for _, d := range snap {
+		dest := d.(v1alpha1.ClusterLogDestination)
+		destNameToObject[dest.Name] = dest
+	}
+
+	return &Generator{logDestinations: destNameToObject}
+}
+
+func (g *Generator) Do(input *go_hook.HookInput) ([]byte, error) {
+	clusterSources, err := g.clusterSources(input)
+	if err != nil {
+		input.LogEntry.Warn(err)
+	}
+
+	namespacedSources, err := g.namespacedSources(input)
+	if err != nil {
+		input.LogEntry.Warn(err)
+	}
+
+	logConfigGenerator := config.NewLogConfigGenerator()
+
+	for _, source := range clusterSources {
+		pipeline, err := config.NewPipelineCluster(logConfigGenerator, g.logDestinations, &source)
+		if err != nil {
+			input.LogEntry.Warn(err)
+			continue
```

Comment:
If we get an error we will skip the pipeline but will keep the rest of the config. Does it valid? We will have some cutted config which can lead to unexpected behavior

Reply:
The pipeline is a valid part of the config (the logical one). As I see it, it is not a big deal because we duplicate sinks, sources, and transforms.

Btw, I think this code is in [master](https://github.com/deckhouse/deckhouse/blob/ecd86c64814c5b7a4bdd13cd2fe3d2ec3c49d1f9/modules/460-log-shipper/hooks/generate_config.go#L239-L257) already.
I just moved it to the dedicated method. And I believe that this code is working like this from the [very beginning (v1.28)](https://github.com/deckhouse/deckhouse/blob/3b69c70aa47dcfe1ed3bc1f360f557aa2273e320/modules/460-log-shipper/hooks/generate_config.go#L199-L219).